### PR TITLE
PromiseKit/Replace synchronization primitive

### PR DESCRIPTION
**Pull request by tonikocjan**
### Replace `DispatchQueue` with `NSLock` as a locking primitive

Based on measurements, `NSLock` is 3x faster, and in the best case, even as much as 10x quicker than `DispatchQueue` using `sync(barrier:)` (`async(barrier:)` is much slower still).

---

## Benchmark 1:

This benchmark spawns different amount of threads, where each performs an operation (either read or write) multiple times. We see linear scalability with both locking primitives; however, `NSLock` scales much better (being around 3x faster).

read
Write

## Benchmark 2:

In this benchmark, we set up a more realistic test scenario that mimics usage in our apps. We performed three tests, a read-intensive test, a write-intensive test, and a mixed test. Following are the results:

### Read-intensive test:

**DispatchQueue:**
Avg: 0.41s
Std: 0.19s

**NSLock**:
Avg: 0.11s
Std: 0.07s

__A ~3.72x improvement.__

### Write-intensive test:

**DispatchQueue**:
Avg: 0.50s
Std: 0.03s

**NSLock**:
Avg: 0.20s
Std: 0.08s

__A ~2.5x improvement.__

### Mixed test:

**DispatchQueue**:
Avg: 4.27s
Std: 2.99s

**NSLock**:
Avg: 0.45s
Std: 0.12s

__A ~9,5x improvement.__

We analyzed the performance of `Mutex` and `RWLock` (read-write lock) too, but they achieve similar performance as `NSLock`, hence `NSLock` was chosen.

---

Additional testing on actual projects is welcomed. :)

### CODE REVIEW

**Review by gentique — APPROVED** [review]
Nice work 👏

File: Sources/PromiseKit/Future.swift [hunk 1]
```diff
@@ -76,3 +88,21 @@ private extension Future {
...
+  }
+  
+  @inlinable
+  @inline(__always)
+  func write(_ work: () -> Void) {
+    lock()
+    defer { unlock() }
+    return work()
+  }
+}
```
Comment (borut-t) [inline Sources/PromiseKit/Future.swift:108]: Can we move this under PromiseKit/Extensions folder instead?

File: Sources/PromiseKit/Promise.swift [hunk 2]
```diff
@@ -873,6 +873,8 @@ public extension Promise where Value == Void {
 }
 
 extension Optional where Wrapped == DispatchQueue {
+  @inlinable
+  @inline(__always)
```
Comment (borut-t) [inline Sources/PromiseKit/Promise.swift:877]: Why using `inlinable` and `inline`?

**Review by borut-t — COMMENTED** [review]
Great work. Have some second thoughts about `inline` attribute. Can you elaborate its usage?

File: Sources/PromiseKit/Promise.swift (hunk 2 above)
Comment (tonikocjan) [reply Sources/PromiseKit/Promise.swift:877]: The way swift has this implemented is a bit confusing. The in-depth discussion can be seen here: https://forums.swift.org/t/when-should-both-inlinable-and-inline-always-be-used/37375/2

File: Sources/PromiseKit/Future.swift (hunk 1 above)
Comment (tonikocjan) [reply Sources/PromiseKit/Future.swift:108]: It's not part of the public API, and I think it should not be.

File: Sources/PromiseKit/Promise.swift (hunk 2 above)
Comment (borut-t) [reply Sources/PromiseKit/Promise.swift:877]: I've read about it [here](https://swiftrocks.com/understanding-inlinable-in-swift.html) and [here](https://lists.swift.org/pipermail/swift-users/Week-of-Mon-20170227/004886.html) as well. They don't recommend it. It's public API but undocumented.

File: Sources/PromiseKit/Future.swift (hunk 1 above)
Comment (borut-t) [reply Sources/PromiseKit/Future.swift:108]: What do you mean? It is already, ust put it in a separate file 🤔

File: Sources/PromiseKit/Promise.swift (hunk 2 above)
Comment (tonikocjan) [reply Sources/PromiseKit/Promise.swift]: Removed @inlinable.

File: Sources/PromiseKit/Future.swift (hunk 1 above)
Comment (tonikocjan) [reply Sources/PromiseKit/Future.swift]: It is not. For it it be public it has to have a `public modifier`.

Comment (borut-t) [reply Sources/PromiseKit/Future.swift]: But if it's internal access, other methods can still see it without exposing the interface right?

**Review by markom-povio — APPROVED** [review]

File: Sources/PromiseKit/Future.swift (hunk 1 above)
Comment (tonikocjan) [reply Sources/PromiseKit/Future.swift]: Decided to mark the extension as `fileprivate` to prevent others using it.

**Review by borut-t — APPROVED** [review]
Great work 🙌

**Pull request merged**